File: AI_cpp/1bfsdfs.cpp
#include <iostream>
#include <queue>
#include <stack>
using namespace std;

int main() {
    int n, m;
    cout << "Enter the number of vertices: ";
    cin >> n;
    cout << "Enter the number of edges: ";
    cin >> m;

    int a[n][n] = {0};  // Initialize the adjacency matrix with zeros

    cout << "\nEnter the vertices between which edges are present:\n";
    int v1, v2;
    for (int i = 0; i < m; i++) {
        cout << "Vertex: ";
        cin >> v1;
        cout << "Vertex: ";
        cin >> v2;
        cout << "Vertex " << v1 << " <---------> Vertex " << v2 << endl;
        a[v1][v2] = 1;
        a[v2][v1] = 1;
    }

    int root;
    cout << "\nEnter the root vertex: ";
    cin >> root;

    // BFS Traversal
    queue<int> q;
    bool visited[n] = {false};  // Initialize all vertices as unvisited
    q.push(root);
    cout << "\nBFS Traversal:\n";
    while (!q.empty()) {
        int x = q.front();
        q.pop();
        if (visited[x]) {
            continue;
        }
        cout << x << " ";
        visited[x] = true;
        for (int i = 0; i < n; i++) {
            if (a[x][i] == 1 && !visited[i]) {
                q.push(i);
            }
        }
    }

    // Reset visited array for DFS traversal
    for (int i = 0; i < n; i++) {
        visited[i] = false;
    }

    // DFS Traversal
    stack<int> s;
    s.push(root);
    cout << "\n\nDFS Traversal:\n";
    while (!s.empty()) {
        int x = s.top();
        s.pop();
        if (visited[x]) {
            continue;
        }
        cout << x << " ";
        visited[x] = true;
        for (int i = 0; i < n; i++) {
            if (a[x][i] == 1 && !visited[i]) {
                s.push(i);
            }
        }
    }

    return 0;
}

File: AI_py/1bfsdfs.py
from queue import Queue
from queue import LifoQueue

n = int(input("Enter the number of vertices: "))
m = int(input("Enter the number of edges: "))

a = [[0] * n for _ in range(n)]  # Initialize the adjacency matrix with zeros

print("\nEnter the vertices between which edges are present:")
for _ in range(m):
    v1 = int(input("Vertex: "))
    v2 = int(input("Vertex: "))
    print(f"Vertex {v1} <---------> Vertex {v2}")
    a[v1][v2] = 1
    a[v2][v1] = 1

root = int(input("\nEnter the root vertex: "))

# BFS Traversal
q = Queue()
visited = [False] * n  # Initialize all vertices as unvisited
q.put(root)
print("\nBFS Traversal:")
while not q.empty():
    x = q.get()
    if visited[x]:
        continue
    print(x, end=" ")
    visited[x] = True
    for i in range(n):
        if a[x][i] == 1 and not visited[i]:
            q.put(i)

# Reset visited array for DFS traversal
visited = [False] * n

# DFS Traversal
s = LifoQueue()
s.put(root)
print("\n\nDFS Traversal:")
while not s.empty():
    x = s.get()
    if visited[x]:
        continue
    print(x, end=" ")
    visited[x] = True
    for i in range(n):
        if a[x][i] == 1 and not visited[i]:
            s.put(i)
